index.js

```diff
@@ -1,154 +1,280 @@
-  if (this.options.inject) {
-    html = this.injectAssetsIntoHtml(html, templateParams);
+  if(typeof source === 'string') {
+    // If the compilation result is already evaluted return it
+    return Promise.resolve(source);
   }
-
-  // Minify the html output
-  if (this.options.minify) {
-    var minify = require('html-minifier').minify;
-    // If `options.minify` is set to true use the default minify options
-    var minifyOptions = _.isObject(this.options.minify) ? this.options.minify : {};
-    html = minify(html, minifyOptions);
+  // Strip the leading 'var '
+  var source = compilationResult.source().substr(3);
+  // Evaluate code and cast to string
+  var newSource;
+  try {
+    newSource = vm.runInThisContext(source);
+  } catch (e) {
+    // Log syntax error
+    var syntaxError = require('syntax-error')(source);
+    var errorMessage = 'Template compilation failed: ' + e + 
+      (syntaxError ? '\n' + syntaxError + '\n\n\n' + source.split('\n').map(function(row, i) {
+        return (1 + i) + '  - ' + row;
+      }).join('\n') : '');
+    compilation.errors.push(new Error(errorMessage));
+    return Promise.reject(e);
   }
+  return typeof newSource === 'string' || typeof newSource === 'function' ?
+    Promise.resolve(newSource) :
+    Promise.reject('The loader "' + this.options.template + '" didn\'t return html.');
+};
 
-  compilation.assets[outputFilename] = {
-    source: function() {
+/**
+ * Html post processing
+ *
+ * Returns a promise
+ */
+HtmlWebpackPlugin.prototype.executeTemplate = function(templateFunction, chunks, assets, compilation) {
+  var self = this;
+  return Promise.resolve()
+    // Template processing
+    .then(function() {
+      var templateParams = {
+        webpack: compilation.getStats().toJson(),
+        webpackConfig: compilation.options,
+        htmlWebpackPlugin: {
+          files: assets,
+          options: self.options,
+        }
+      };
+      var html = '';
+      try {
+        html = templateFunction(templateParams);
+      } catch (e) {
+        compilation.errors.push(new Error('Template execution failed: ' + e));
+        return Promise.reject(e);
+      }
       return html;
-    },
-    size: function() {
-      return html.length;
-    }
-  };
+    });
+};
+
+/**
+ * Html post processing
+ *
+ * Returns a promise
+ */
+HtmlWebpackPlugin.prototype.postProcessHtml = function(html, assets) {
+  var self = this;
+  return Promise.resolve()
+    // Inject
+    .then(function() {
+      if (self.options.inject) {
+        return self.injectAssetsIntoHtml(html, assets);
+      } else {
+        return html;
+      }
+    })
+    // Minify
+    .then(function(html) {
+      if (self.options.minify) {
+        var minify = require('html-minifier').minify;
+        // If `options.minify` is set to true use the default minify options
+        var minifyOptions = _.isObject(self.options.minify) ? self.options.minify : {};
+        try {
+          minify(html, minifyOptions);
```

Comment:
This here destroys `minify` option. Either return the result from `minify`, or assign it to `html`